### ISSUE TITLE
I think the ANC source code might just be available

### ISSUE THREAD

**Issue by WebWire-NL** [pine64/OpenPineBuds#25]
If i'm right the ANC source code might just be available on this site:
https://blog.csdn.net/qq_34990604/article/details/119966060
Seems they also provide some sort of SDK you need to subscribe to get any info with some weird accounts but i think the source will be just there, seems ANC is integrated in the chip if i'm right.

**Comment by Ralim** [comment]
Yeah I haven't bee able to gain access to this code sadly; (cant create csdn account and put money on it).
I think its software that manages hardware mixing filters or some such setup.
(Not sure; guessing).

**Comment by WebWire-NL** [comment]
Wish we had known earlier, the BES2500 SDK was just available for download not a long time ago :)
https://programmersought.com/article/27637776906/
But still on the site i mentioned there's a lot of information which might be usefull to enable the anc function. We might also be able to just extract the app from the firmware and inject it into the firmware again. But like i said.... still not concentrated enough to digg in...

**Comment by WebWire-NL** [comment]
Found something else!
https://github.com/ZhiqingLi/Sdk_Refresh/tree/master/BES2000_dueros/apps/anc

**Comment by WebWire-NL** [comment]
Just downloaded the Pine64 SDK and it seems it's already there... so we only need to enable the right values in the config right? as described on the blog....

**Comment by happyme531** [comment]
> Yeah I haven't bee able to gain access to this code sadly; (cant create csdn account and put money on it). I think its software that manages hardware mixing filters or some such setup. (Not sure; guessing).

There are plenty of 3rd party downloading sites, such as https://ziquyun.com/main/csdn?zdurl=https%3A%2F%2Fblog.csdn.net%2Fzhanghuaishu0%2Fcategory_10150640.html&zdn=true
(This is not an ad, Chinese only, hold your translator tight)
And you do not need to register an account for it, also much cheaper than the offical site.

**Comment by Ralim** [comment]
> I'll stop :) doesn't look too hard as the configs are already there....

You say this; but the build we are using _has_ ANC turned on.
When you "enable" it via button press you get _nothing_.
The logs state its enabled but no change occurs.

Is started doing a decompilation of the binary blob and found some of the values it checks against and so forced the firmware into those states and got something (screeching noise that hurts like crazy if its near your ear). So it does _something_ but we have no calibration values, or really any idea of what is actually required. It could be trivial but so far all the docs I have been able to read I have poked around with havent resulting in working ANC.

Config values from the blog did _not_ do anything for me (so far in testing).

I saw we dont really have it as we only have the binary blob; which was compiled with optimisations so we only have _one_ code path in there (there are lots missing). So we have _something_ but I wouldn't call it having it all.


I'll see about the 3rd party sites; hadnt seen these before.

**Comment by WebWire-NL** [comment]
Maybe this source can also help, https://github.com/ZhiqingLi/Sdk_Refresh/search?p=1&q=anc
I see the audioflinger thingie mentioned in a lot of posts....

**Comment by WebWire-NL** [comment]
@Ralim is it possible to make an ANC branch so we can see what you've tried already so we don't have to reinvent the wheel? :)
Also maybe share some tools which you used to decompile/analyze the firmware? As i read in the blogs they added a specific binary to the firmware to test the ANC. Still don't have enough concentration/motivation to digg into the source :( but try to do whatever research i can.

**Comment by WebWire-NL** [comment]
Reopening this with something concrete, since the archives turned out to be less opaque than we assumed.

@Ralim's summary above ("we only have the binary blob, which was compiled with optimisations so we only have one code path") is right for the *production firmware dumps* - those are stripped. The SDK's prebuilt archives are a different story: they were built with `-g` and still carry DWARF.

- `services/multimedia/lib/best2300p_libmultimedia_cp_anc.a` - 162 objects, 332 source files named in DWARF, all of them absent from the tree, GCC 9.2.1 `-g -O2`. This is the archive with the ANC/DSP code (adj_mc/ANC processing, the mic denoise chain, HW/SW IIR+FIR EQ).
- `services/anc_spp_tool/lib/libanc_spp_tool.a` - small (2 objects, 9 KB) and also with DWARF: the ANC over-SPP tuning path, i.e. the thing a tuning host talks to.
- The archives *without* debug info are the IBRT core/UI/OTA ones, LHDC and intersyshci. So "no calibration values, no idea what is required" is a problem of *values and tuning*, not of the binary being opaque.

The DWARF turns back into declarations: 593 ANC/DSP prototypes with return types, argument types and source line numbers, and 31 generated headers - the ANC entry points, the filter/EQ stages and their parameters, by name instead of by guesswork.
- https://github.com/WebWire-NL/openpinebuds-integration/blob/main/docs/anc-prototypes.md
- https://github.com/WebWire-NL/openpinebuds-integration/tree/main/include/anc_reconstructed

Boundary, stated explicitly: interface metadata only - no vendor source, no decompiled code and no binaries are published. The implementations are not public anywhere (checked the large public BES2300/2600/2700 SDK trees that do carry the headers).

Fork work that is closer to "having it" than the blob: `nnonickreal/openqore-sdk` (BES2300P port that wires ANC/transparency through the `aud` NV section, 37 commits ahead), `SilentBob347/OpenPineBuds` (`anc-tuning`, `anc-testing`), `Haxk20/OpenPineBuds` (`anc-testing`), plus upstream `dev_tools/anc_decoder`. Survey and replayable patches: https://github.com/WebWire-NL/openpinebuds-integration

Analysis by @WebWire-NL. Useful in #41 too - reuse it there with a link back if so.
